crates/storage/provider/src/post_state/account.rs

```diff
@@ -18,10 +18,26 @@ pub struct AccountChanges {
 impl AccountChanges {
     /// Insert account change at specified block number. The value is **not** updated if it already
     /// exists.
-    pub fn insert(&mut self, block: BlockNumber, address: Address, account: Option<Account>) {
-        if let Entry::Vacant(entry) = self.inner.entry(block).or_default().entry(address) {
-            self.size += 1;
-            entry.insert(account);
+    pub fn insert(
+        &mut self,
+        block: BlockNumber,
+        address: Address,
+        old: Option<Account>,
+        new: Option<Account>,
+    ) {
+        match self.inner.entry(block).or_default().entry(address) {
+            Entry::Vacant(entry) => {
+                self.size += 1;
+                entry.insert(old);
+            }
+            Entry::Occupied(entry) => {
+                // If the account state is the same before and after this block, collapse the state
+                // changes.
+                if entry.get() == &new {
+                    entry.remove();
+                    self.size -= 1;
+                }
+            }
```

Comment:
<s>Can you explain a bit what you mean by collapse? Like combine, right? If so, this looks a bit weird given we're removing the old, and we're not inserting the new account? So there's no change?</s>

I see this means that in a block-level changeset, there's no need to insert a change if the before/after is the same. Would update the doc comment to reflect that, i.e. if there was a A->A'->A changeset, we end up storing no changeset.